### PR TITLE
Map docker images to python version

### PR DESCRIPTION
**Pull request by GuyAfik**
[related issue](https://jira-hq.paloaltonetworks.local/browse/CIAC-7277)

### CODE REVIEW

File: update-docker-repo-info.py [hunk 1]
```diff
@@ -22,6 +22,11 @@
 VERIFY_SSL = True
 
 DOCKERFILES_DIR = os.path.abspath(os.getenv('DOCKERFILES_DIR', '.dockerfiles'))
+try:
+    with open("docker_images_metadata.json", "r") as f:
+        DOCKER_IMAGES_METADATA = json.load(f)
+except Exception:
+    DOCKER_IMAGES_METADATA = {}
```
Comment (kobymeir) [inline update-docker-repo-info.py]: perhaps worth checking only json parsing exception. and printing if the file isn't there or we failed to load the json.

File: docker_images_metadata.json [hunk 2]
```diff
@@ -0,0 +1,1486 @@
...
+         "1.0.0.49364":{
+            "python_version":"3.10.10"
+         }
+      },
+      "archer":{
+         "1.0.0.270":{
+            "python_version":"3.7.3"
+         }
+      },
+      "powershell-teams":{
```
Comment (kobymeir) [inline docker_images_metadata.json]: why it's empty here? if it doesn't have any tags, just don't keep it

File: update-docker-repo-info.py [hunk 3]
```diff
@@ -92,6 +97,12 @@ def get_os_release(image_name):
     return res.stdout.splitlines()
 
 
+def get_python_version(docker_info: str) -> str:
+    if python_version := re.search(r'PYTHON_VERSION=(\d+\.\d+\.\d+)', docker_info):
+        return python_version.group(1)
+    return ''
```
Comment (kobymeir) [inline update-docker-repo-info.py]:
```suggestion
    return python_version[1] if (python_version := re.search(r'PYTHON_VERSION=(\d+\.\d+\.\d+)', docker_info)) else ''
```

**Review by kobymeir — CHANGES_REQUESTED** [review]

File: update-docker-repo-info.py (hunk 1 above)
Comment (GuyAfik) [reply update-docker-repo-info.py]: done

File: docker_images_metadata.json (hunk 2 above)
Comment (GuyAfik) [reply docker_images_metadata.json]: removed them

File: update-docker-repo-info.py [hunk 4]
```diff
@@ -105,7 +117,16 @@ def inspect_image(image_name, out_file):
 - Labels:{{{{ range $key, $value := .ContainerConfig.Labels }}}}{{{{ "\\n" }}}}  - `{{{{ $key }}}}:{{{{ $value }}}}`{{{{ end }}}}
 '''
     docker_info = subprocess.check_output(["docker", "inspect", "-f", inspect_format, image_name], text=True)
+
     out_file.write(docker_info)
+
+    # get python version and add it to the docker images metadata file
+    if not DOCKER_IMAGES_METADATA.get(image_name) and (python_version := get_python_version(docker_info)):
```
Comment (JudahSchwartz) [inline update-docker-repo-info.py]: need to add when we have image name

File: update-docker-repo-info.py [hunk 5]
```diff
@@ -105,7 +121,27 @@ def inspect_image(image_name, out_file):
 - Labels:{{{{ range $key, $value := .ContainerConfig.Labels }}}}{{{{ "\\n" }}}}  - `{{{{ $key }}}}:{{{{ $value }}}}`{{{{ end }}}}
 '''
     docker_info = subprocess.check_output(["docker", "inspect", "-f", inspect_format, image_name], text=True)
+
     out_file.write(docker_info)
+
+    # get python version and add it to the docker images metadata file
+    try:
+        docker_name, tag = image_name.replace("demisto/", "").split(":")
```
Comment (kobymeir) [inline update-docker-repo-info.py]: why not regex?

**Review by kobymeir — CHANGES_REQUESTED** [review]

File: update-docker-repo-info.py (hunk 5 above)
Comment (GuyAfik) [reply update-docker-repo-info.py]: not a specific reason, changed to regex

File: update-docker-repo-info.py [hunk 6]
```diff
@@ -398,13 +448,15 @@ def checkout_dockerfiles_repo():
...
 
+
 def main():
     parser = argparse.ArgumentParser(description='Fetch docker repo info. Will fetch the docker image and then generate license info',
                                      formatter_class=argparse.ArgumentDefaultsHelpFormatter)
     parser.add_argument("docker_image", help="The docker image name to use (ie: demisto/python). Optional." +
                         "If not specified will scan all images in the demisto organization", nargs="?")
     parser.add_argument("--force", help="Force refetch even if license data already exists", action='store_true')
     parser.add_argument("--no-verify-ssl", help="Don't verify ssl certs for requests (for testing behind corp firewall)", action='store_true')
+    update_docker_files_metadata_json()
```
Comment (JudahSchwartz) [inline update-docker-repo-info.py]: why do we call this function here?

File: update-docker-repo-info.py [hunk 7]
```diff
@@ -15,13 +15,25 @@
...
 import time
+import codecs
 
 
 assert sys.version_info >= (3, 9), "Script compatible with python 3.9 and higher only"
 
 VERIFY_SSL = True
 
 DOCKERFILES_DIR = os.path.abspath(os.getenv('DOCKERFILES_DIR', '.dockerfiles'))
+DOCKER_IMAGES_METADATA = "docker_images_metadata.json"
```
Comment (JudahSchwartz) [inline update-docker-repo-info.py:26]: not good practice to have a variable be a file name, and then file contents. Lets split to different variables

File: update-docker-repo-info.py [hunk 8]
```diff
@@ -15,13 +15,25 @@
...
+DOCKER_IMAGE_REGEX_PATTERN = r'^demisto/([^\s:]+):(\d+(\.\d+)*)$'
+
+
+
+try:
+    with codecs.open(DOCKER_IMAGES_METADATA, encoding="utf-8-sig") as f:
+        DOCKER_IMAGES_METADATA = json.load(f)
+except json.JSONDecodeError:
+    print(f'Could not load {DOCKER_IMAGES_METADATA}')
```
Comment (JudahSchwartz) [inline update-docker-repo-info.py]: we should probably abort this whole operation if this is what happened. We dont want to overwrite with an empty file and lose everything

File: update-docker-repo-info.py [hunk 9]
```diff
@@ -389,6 +434,11 @@ def generate_csv():
                                 value.get('author'), value.get('summary'), ", ".join(value.get('docker_images'))])
 
 
+def update_docker_files_metadata_json():
```
Comment (JudahSchwartz) [inline update-docker-repo-info.py]: lets rename with "save to file" or something that its clear we're flushing to the file

File: update-docker-repo-info.py [hunk 10]
```diff
@@ -92,6 +104,34 @@ def get_os_release(image_name):
     return res.stdout.splitlines()
 
 
+def get_python_version(docker_info: str) -> str:
```
Comment (JudahSchwartz) [inline update-docker-repo-info.py:106]: not related to this pr, but I noticed this env var r'PYTHON_VERSION doesnt exist for the ubi image. Maybe we should add it in the Dockerfile manually? Lets discuss

File: update-docker-repo-info.py [hunk 11]
```diff
@@ -92,6 +104,34 @@ def get_os_release(image_name):
...
+    return python_version.group(1) if (
+        python_version := re.search(r'PYTHON_VERSION=(\d+\.\d+\.\d+)', docker_info)
+    ) else ''
+
+
+def add_python_version_for_dockerfiles_metadata(image_name: str, docker_info: str):
+    try:
+        if match := re.match(DOCKER_IMAGE_REGEX_PATTERN, image_name):
```
Comment (JudahSchwartz) [inline update-docker-repo-info.py:114]: is there a legitimate case why this would fail?

File: update-docker-repo-info.py (hunk 7 above)
Comment (GuyAfik) [reply update-docker-repo-info.py:26]: i didn't even notice i did it, nice catch, its a mistake

File: update-docker-repo-info.py (hunk 11 above)
Comment (GuyAfik) [reply update-docker-repo-info.py:114]: its better to handle the case that it fails so we would know there is a problem in the future, i don't want to assume that every input is valid.

File: update-docker-repo-info.py (hunk 6 above)
Comment (GuyAfik) [reply update-docker-repo-info.py]: it was just a test i made, will remove it

File: update-docker-repo-info.py (hunk 8 above)
Comment (GuyAfik) [reply update-docker-repo-info.py]: agreed

File: update-docker-repo-info.py (hunk 10 above)
Comment (GuyAfik) [reply update-docker-repo-info.py:106]: https://github.com/demisto/dockerfiles/pull/19210

Comment (GuyAfik) [reply update-docker-repo-info.py:106]: merged

**Review by JudahSchwartz — APPROVED** [review]
looks good

**Review by kobymeir — APPROVED** [review]

**Pull request merged**